tools/testing/e2e/iree-e2e-attention-test.cc

```diff
@@ -0,0 +1,476 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "iree/base/api.h"
+#include "iree/base/internal/cpu.h"
+#include "iree/base/internal/flags.h"
+#include "iree/base/internal/math.h"
+#include "iree/base/internal/path.h"
+#include "iree/hal/api.h"
+#include "iree/modules/hal/module.h"
+#include "iree/tooling/context_util.h"
+#include "iree/tooling/device_util.h"
+#include "iree/vm/api.h"
+#include "iree/vm/native_module_cc.h"
+#include "tools/testing/e2e/test_utils.h"
+
+//===----------------------------------------------------------------------===//
+// Reference Attention
+//===----------------------------------------------------------------------===//
+
+// Helper for reference_attention.
+// Function to allocate and initialize tensors
+float* allocate_tensor(int dim1, int dim2, int dim3) {
+  float* tensor = (float*)malloc(dim1 * dim2 * dim3 * sizeof(float));
+  for (int i = 0; i < dim1 * dim2 * dim3; ++i) {
+    tensor[i] = 0.0f;
+  }
+  return tensor;
+}
+
+// Function to free allocated tensors
+void free_tensor(float* tensor) {
+  if (tensor != nullptr) free(tensor);
+}
+
+// Function to calculate 1D index for a 3D array
+int index_3d(int i, int j, int k, int dim2, int dim3) {
+  return i * dim2 * dim3 + j * dim3 + k;
+}
+
+static void reference_attention_f32_f32_f32_f32(
+    iree_hal_dim_t M, iree_hal_dim_t K1, iree_hal_dim_t K2, iree_hal_dim_t N,
+    iree_hal_dim_t B, const float* query_data, const float* key_data,
+    const float* value_data, float* result_data, iree_hal_dim_t b,
+    float* Attention) {
+  // Compute Q * K^T
+  for (int m = 0; m < M; ++m) {
+    for (int k2 = 0; k2 < K2; ++k2) {
+      float sum = 0.0;
+      for (int k1 = 0; k1 < K1; ++k1) {
+        int q_idx = index_3d(b, m, k1, M, K1);
+        int k_idx = index_3d(b, k2, k1, K2, K1);
+
+        sum += query_data[q_idx] * key_data[k_idx];
+      }
+      int att_idx = index_3d(0, m, k2, M, K2);
+      Attention[att_idx] = sum / sqrt(K1);  // Scale by sqrt(K1)
+    }
+  }
+
+  // Compute softmax on Attention
+  for (int m = 0; m < M; ++m) {
```

Comment:
Please use the max implementation for softmax. This might cause numerical instability. Can read https://stackoverflow.com/questions/34968722/how-to-implement-the-softmax-function-in-python

Reply:
Thanks, Prashant, I changed that.